include/boost/beast/core/buffers_generator.hpp

```diff
@@ -0,0 +1,229 @@
+    T, detail::void_t<decltype(
+        typename T::const_buffers_type(
+            std::declval<T&>().prepare(
+                std::declval<error_code&>())),
+        std::declval<T&>().consume(
+            std::size_t{})
+    )>> : std::true_type
+{
+};
+#endif
+
+//----------------------------------------------------------
+
+namespace detail {
+
+template <
+    class AsyncWriteStream,
+    class BuffersGenerator>
+struct write_buffers_generator_op
+    : boost::asio::coroutine
+{
+    write_buffers_generator_op(AsyncWriteStream& stream,
+                               BuffersGenerator generator)
+        : stream_(stream)
+        , generator_(std::move(generator))
+    {
+    }
+
+    template <class Self>
+    void operator()(Self& self,
+                    error_code ec = {},
+                    std::size_t n = 0)
+    {
+        BOOST_ASIO_CORO_REENTER(*this)
+        {
+            for(;;)
+            {
+                BOOST_ASIO_CORO_YIELD
+                {
+                    auto cb = generator_.prepare(ec);
+                    if (ec)
+                        goto complete;
+                    if (boost::beast::buffer_bytes(cb) == 0)
+                        goto complete;
+                    stream_.async_write_some(
+                        cb, std::move(self));
+                }
+                generator_.consume(n);
+
+                total_ += n;
+            }
+
+        complete:
+            self.complete(ec, total_);
+        }
+    }
+
+private:
+    AsyncWriteStream& stream_;
+    BuffersGenerator generator_;
+    std::size_t total_ = 0;
+};
+
+} // detail
+
+//----------------------------------------------------------
+
+template <
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<typename std::decay<
+            BuffersGenerator>::type>::value>::type* =
+        nullptr
+#endif
+    >
+size_t
+write(SyncWriteStream& stream,
+      BuffersGenerator&& generator,
+      beast::error_code& ec)
+{
+    static_assert(
+        is_sync_write_stream<SyncWriteStream>::value,
+        "SyncWriteStream type requirements not met");
+
+    ec.clear();
+    size_t total = 0;
+    for (;;) {
+        auto cb = generator.prepare(ec);
+        if (ec)
+            break;
+        if (boost::beast::buffer_bytes(cb) == 0)
+            break;
+
+        size_t n = net::write(stream, cb, ec);
+
+        if (ec)
+            break;
+
+        generator.consume(n);
+        total += n;
+    }
+
+    return total;
+}
+
+//----------------------------------------------------------
+
+template <
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<typename std::decay<
+            BuffersGenerator>::type>::value>::type* =
+        nullptr
+#endif
+    >
+std::size_t
+write(SyncWriteStream& stream, BuffersGenerator&& generator)
```

Comment:
This needs to go in `<boost/beast/core/impl/buffers_generator.hpp>` so we don't force the doc toolchain to parse it needlessly.

Reply:
Done